// components/Home/index.js
import 'components/Home/Home.css'
import React, {Component} from 'react'
import HeroImage from 'components/elements/HeroImage'
import SearchBar from 'components/elements/SearchBar'
import Spinner from 'components/elements/Spinner'
import FourColGrid from 'components/elements/FourColGrid'
import MovieThumd from 'components/elements/MovieThumb'
import LoadMoreBtn from 'components/elements/LoadMoreBtn'
import {API_KEY, API_URL, IMAGE_BASE_URL, BACKDROP_SIZE, POSTER_SIZE} from "config";
import no_img from 'images/no_image.jpg'
import axios from 'axios'
import {Link} from "react-router-dom";


export default class Home extends Component{

    state = {
        movies: [],
        heroImage: null,
        loading: false,
        currentPage: 0,
        totalPage: 0,
        searchTerm: '',
        iterator: 16
    }

    componentDidMount() {
        this.setState({loading: true})
        const endpoint = `${API_URL}movie/popular?api_key=${API_KEY}&language=en-US&page=1`
        this.fetchItems(endpoint)
    }

    searchItems = (searchTerm) => {

        let endpoint = ''

        this.setState({
            movies: [],
            loading: true,
            searchTerm
        })

        if(searchTerm === ''){
            endpoint = `${API_URL}movie/popular?api_key=${API_KEY}&language=en-US&page=1`
        } else {
            endpoint = `${API_URL}search/movie?api_key=${API_KEY}&language=en-US&query=${searchTerm}`
        }

        this.fetchItems(endpoint)
    }

    loadMoreItems = () => {
        let endpoint = ''

        this.setState({loading: true})

        if(this.state.searchTerm === ''){
            endpoint = `${API_URL}movie/popular?api_key=${API_KEY}&language=en-US&page=${this.state.currentPage + 1}`
        } else {
            endpoint = `${API_URL}search/movie?api_key=${API_KEY}&language=en-US&query=${this.state.searchTerm}&page=${this.state.currentPage + 1}`
        }

        this.fetchItems(endpoint)
    }

    fetchItems = endpoint => {

        try {
            axios.get(endpoint)
                .then(result => {

                    this.setState({
                        movies: [...this.state.movies, ...result.data.results],
                        heroImage: this.state.heroImage || result.data.results[this.state.iterator],
                        loading: false,
                        currentPage: result.data.page,
                        totalPage: result.data.total_pages
                    })

                    console.log(result)
                })
        }catch(e){
            console.log(e)
        }
    }

    render(){
        return(
            <div className='rmdb-home'>
                {this.state.heroImage ?
                <div>
                        <HeroImage
                            image={`${IMAGE_BASE_URL}${BACKDROP_SIZE}${this.state.heroImage.backdrop_path}`}
                            title={this.state.heroImage.original_title}
                            text={this.state.heroImage.overview}
                        />
                    <SearchBar callback={this.searchItems} />
                </div> : null }
                <div className='rmdb-home-grid'>
                    <FourColGrid
                    header={this.state.searchTerm ? 'Результаты поиска' : 'Популярные картины'}
                    loading={this.state.loading}
                    >
                        {this.state.movies.map((element, i) => {
                            return <MovieThumd
                            key={i}
                            clicable={true}
                            image={element.poster_path ? `${IMAGE_BASE_URL}${POSTER_SIZE}${element.poster_path}` : no_img}
                            movieId={element.id}
                            movieName={element.original_title}
                            />
                        })}
                    </FourColGrid>
                    {this.state.loading ? <Spinner /> : null}
                    {
                        this.state.currentPage <= this.state.totalPage && !this.state.loading ?
                        <LoadMoreBtn text='Еще фильмы' onClick={this.loadMoreItems} />
                        : null
                    }
                </div>
            </div>
        )
    }
}
