// components/Movie/index.js
import React, {Component} from 'react'
import {API_KEY, API_URL} from "config";
import Navigation from 'components/elements/Navigation'
import MovieInfo from 'components/elements/MovieInfo'
import MovieInfoBar from 'components/elements/MovieInfoBar'
import FourColGrid from 'components/elements/FourColGrid'
import Actor from 'components/elements/Actor'
import Spinner from 'components/elements/Spinner'
import 'components/Movie/Movie.css'
import axios from 'axios'


export default class Auth extends Component{

    state = {
        movie: null,
        actors: null,
        directors: [],
        loading: false
    }

    componentDidMount() {
        this.setState({loading: true})

        const endpoint = `${API_URL}movie/${this.props.match.params.movieId}?api_key=${API_KEY}&language=en-US`

        this.fetchItems(endpoint)
    }

    fetchItems = endpoint => {
        axios.get(endpoint)
            .then(res => {
                console.log(res.data)

                if(res.status_code){
                    this.setState({loading: false})
                } else {
                    this.setState({
                        movie: res.data
                    }, () => {
                        const endpoint = `${API_URL}movie/${this.props.match.params.movieId}/credits?api_key=${API_KEY}&language=en-US`
                        axios.get(endpoint)
                            .then(res => {
                                const directors = res.data.crew.filter(member => member.job === 'Director')

                                this.setState({
                                    actors: res.data.cast,
                                    directors,
                                    loading: false
                                })

                            })
                    })
                }

            })
            .catch(e => console.error(e))
    }

    render(){
        return(
            <div className='rmdb-movie'>
                <Navigation />
                <MovieInfo />
                <MovieInfoBar />
                {/*<FourColGrid />*/}
                <Spinner />
            </div>
        )
    }
}
